python/lsst/obs/base/ingest.py

```diff
@@ -920,6 +920,69 @@ def run(self, files, *, pool: Optional[Pool] = None, processes: int = 1, run: Op
             self.log.info("Exposure %s:%s ingested successfully",
                           exposure.record.instrument, exposure.record.obs_id)
 
+        return refs, bad_files, n_exposures, n_exposures_failed, n_ingests_failed
+
+    @timeMethod
+    def run(self, files, *, pool: Optional[Pool] = None, processes: int = 1, run: Optional[str] = None,
+            file_filter: Union[str, re.Pattern] = r"\.fit[s]?\b", group_files: bool = True):
+        """Ingest files into a Butler data repository.
+
+        This creates any new exposure or visit Dimension entries needed to
+        identify the ingested files, creates new Dataset entries in the
+        Registry and finally ingests the files themselves into the Datastore.
+        Any needed instrument, detector, and physical_filter Dimension entries
+        must exist in the Registry before `run` is called.
+
+        Parameters
+        ----------
+        files : iterable over `ButlerURI`, `str` or path-like objects
+            Paths to the files to be ingested.  Can refer to directories.
+            Will be made absolute if they are not already.
+        pool : `multiprocessing.Pool`, optional
+            If not `None`, a process pool with which to parallelize some
+            operations.
+        processes : `int`, optional
+            The number of processes to use.  Ignored if ``pool`` is not `None`.
+        run : `str`, optional
+            Name of a RUN-type collection to write to, overriding
+            the default derived from the instrument name.
+        file_filter : `str` or `re.Pattern`, optional
+            Pattern to use to discover files to ingest within directories.
+            The default is to search for FITS files.
+        group_files : `bool`, optional
+            Group files by directory if they have been discovered in
+            directories. Will not affect files explicitly provided.
```

Comment:
I wrote something like this for the DC2-specialized ingest wrapper I wrote to setup the shared repos, and I'm happy to see that kind of logic move down into `RawIngestTask` itself.  But for HSC, where the raws tend to be in big, flat, many-exposure directories (but one can read the exposure and detector IDs from the filename), that didn't work - so I instead I used an interface where the caller provided a per-instrument callback to search within a directory, returning groups of files that all belong to the same exposure, and then called `RawIngestTask.run` on each group at a time.  That's all still compatible with your changes here, but it gets me thinking about the overall structure of these interfaces.  I think we've got three different kinds of high-level usage of this task, and maybe we should have different methods (in addition to or instead of `run`) to satisfy them:
 - Given a bag of totally unstructured files, group them into exposures and ingest each exposure at a time.  We have to read all of the metadata before we start ingesting in order to do the grouping, so this doesn't scale, but it's otherwise the most automatic, user-friendly option.  Needs to guess instruments to be totally safe.
 - Ingest a single file, regardless of whether it is the first in its exposure or not.
 - Given a directory tree with well-defined structure for a single instrument, ingest files by exposure according to a per-instrument rule (callback, subclass override, etc.) for that tree structure.  This lets us take a lot of shortcuts (e.g. if per-instrument rules also let us extract data IDs for most raws from the tree structure), and can potentially be much faster, and it scales much better because we can read metadata and then ingest one exposure at a time.  This may be the only mode that needs to be able to produce/read JSON sidecar and index files.  This mode is also much more parallelizable, because we can subdivide the tree early and then do everything entirely independently.

I think we wrote a task designed for the first two cases, and are finding that we care more about the last one, especially when it comes to performance, resume-ability, and careful error-handling, so now supporting the flexibility demanded by the first case in particular is making those things harder.  Now that I've written all of that out, I'm not sure this should even be just one `Task` anymore, and changing that is pretty clearly beyond the scope of this ticket - if it's even worth doing.  But I think it's worth thinking about, if we want one `Task` or a command-line tool derived from it to be able to handle the third case for multiple instruments.

Reply:
Extracting dataIds is not sufficient for ingesting exposures because we are really ingesting `ObservationInfo` in order to populate the exposure record.

We could imagine something where you supply a file matching regex where the subgroups of the regex are detector/exposure/instrument (say) and we look at the URIs and group them by regex, reading one file only and assuming that `obsInfo.detector` will be the only thing that changes per exposure file.  The alternative is to create the `_index.json` using this knowledge directly (I think Eric Charles may be thinking of doing that for BOT data). K-T was pondering a situation where we write index/sidecar files into a parallel tree so that the scanning code doesn't have to find and discard a bunch of files (especially in S3).

Reply:
(and when I say group by exposure/instrument -- they don't have to be the formal exposure/instrument names, they only have to be unique for grouping -- the real values will come from the header)

Reply:
> We could imagine something where you supply a file matching regex where the subgroups of the regex are detector/exposure/instrument (say) and we look at the URIs and group them by regex, reading one file only and assuming that obsInfo.detector will be the only thing that changes per exposure file.

Yup, this is exactly what I was talking about doing for HSC.  And now I realize we're having the same conversation we had a few days ago, where your version is having the per-instrument code writes those index files efficiently, and then having the ingest task read those.  That works for me.